jaxopt/_src/implicit_diff.py

```diff
@@ -120,29 +120,124 @@ def root_jvp(optimality_fun: Callable,
-    # For init_params, we return None.
-    return (None,) + vjps
+  mapping = [(was_kwarg, ref) for was_kwarg, ref, _ in ba.args]
 
-  wrapped_solver_fun = jax.custom_vjp(solver_fun)
-  wrapped_solver_fun.defvjp(solver_fun_fwd, solver_fun_bwd)
+  def map_back(out_args):
+    src_args = [None] * len(args)
+    src_kwargs = {}
+    for (was_kwarg, ref), out_arg in zip(mapping, out_args):
+      if was_kwarg:
+        src_kwargs[ref] = out_arg
+      else:
+        src_args[ref] = out_arg
+    return src_args, src_kwargs
+
+  out_args = tuple(v for _, _, v in ba.args)
+  out_kwargs = {k: v for k, (_, _, v) in ba.kwargs.items()}
+  return out_args, out_kwargs, map_back
+
+
+def _custom_root(solver_fun, optimality_fun, solve, has_aux):
+  # When caling through `jax.custom_vjp`, jax attempts to resolve all
+  # arguments passed by keyword to positions (this is in order to
+  # match against a `nondiff_argnums` parameter that we do not use
+  # here). It does so by resolving them according to the custom_jvp'ed
+  # function's signature. It disallows functions defined with a
+  # catch-all `**kwargs` expression, since their signature cannot
+  # always resolve all keyword arguments to positions.
+  #
+  # We can loosen the constraint on the signature of `solver_fun` so
+  # long as we resolve keywords to positions ourselves. We can do so
+  # just in time, by flattening the `kwargs` dict (respecting its
+  # iteration order) and supplying `custom_vjp` with a
+  # positional-argument-only function. We then explicitly coordinate
+  # flattening and unflattening around the `custom_vjp` boundary.
+  #
+  # Once we make it past the `custom_vjp` boundary, we do some more
+  # work to align arguments with the signature of `optimality_fun`.
+
+  solver_fun_signature = inspect.signature(solver_fun)
+  optimality_signature = inspect.signature(optimality_fun)
+
+  def make_custom_vjp_solver_fun(solver_fun, kwarg_keys):
+    def solver_fun_fwd(*flat_args):
+      args, kwargs = _extract_kwargs(kwarg_keys, flat_args)
+      res = solver_fun(*args, **kwargs)
+      return res, (res, flat_args)
+
+    def solver_fun_bwd(tup, cotangent):
+      res, flat_args = tup
+      args, kwargs = _extract_kwargs(kwarg_keys, flat_args)
+
+      # solver_fun can return auxiliary data if has_aux = True.
+      if has_aux:
+        cotangent = cotangent[0]
+        sol = res[0]
+      else:
+        sol = res
+
+      ba_args, ba_kwargs, map_back = _signature_bind_and_match(
+          optimality_signature, *args, **kwargs)
+      if ba_kwargs:
+        raise TypeError(
+            "keyword arguments to solver_fun could not be resolved to "
+            f"positional arguments based on the signature {signature}. This "
+            "can happen under custom_root if optimality_fun takes catch-all "
+            "**kwargs, or under custom_fixed_point if fixed_point_fun takes "
+            "catch-all **kwargs, both of which are currently unsupported.")
```

Comment:
This error message is rather specific. We may want to catch and re-raise it at higher levels of the library with a more helpful message.